src/java.base/share/classes/com/sun/crypto/provider/CipherBlockChaining.java

```diff
@@ -199,7 +211,15 @@ int decrypt(byte[] cipher, int cipherOffset, int cipherLen,
         ArrayUtil.blockSizeCheck(cipherLen, blockSize);
         ArrayUtil.nullAndBoundsCheck(cipher, cipherOffset, cipherLen);
         ArrayUtil.nullAndBoundsCheck(plain, plainOffset, cipherLen);
-        return implDecrypt(cipher, cipherOffset, cipherLen, plain, plainOffset);
+        int processed = 0;
+        for (;  cipherLen > chunkSize; cipherOffset += chunkSize,
+            plainOffset += chunkSize, cipherLen -= chunkSize) {
+            processed +=
+               implDecrypt(cipher, cipherOffset, chunkSize, plain, plainOffset);
+        }
+        processed +=
+            implDecrypt(cipher, cipherOffset, cipherLen, plain, plainOffset);
+        return processed;
```

Comment:
The `for` loop logic is the same for encrypt and decrypt operations, only different positioning of the arguments. How about creating a helper method `chunkOperation` that would take one additional encrypt/decrypt boolean argument based on which it would do either encrypt or decrypt operation.

Reply:
Given this is a performance change, I'm fine with leaving it as is.  Jumping to a helper method with an encrypt/decrypt conditional check for every crypto op will costs performance.  This is a case where more efficient code is more verbose syntax.

Reply:
@ascarpino But a method call should be very cheap, we are adding a bunch of extra implEncrypt/implDecrypt calls here already. Besides, compiler/hotspot will optimize that call if needed.

Reply:
I don't think it matters either way performance-wise, or from any other point of view in this case, but as a rule of thumb, I think for readability/maintainability it is worth to give up a bit of code size (especially if that is only source code size, since the compiler would duplicate the runtime code anyways) and/or performance. Of course, it is hard to decide which version is more readable/maintainable. For me, in this case, the source code duplication seems to be the better solution, I would not write a helper function for a 3-line for loop. I have spent many hours of my life trying to figure out whether I brake something if I make a little change in a function that appears on multiple code paths...

Reply:
I see, I agree it will add to the learning curve for somebody new looking at the code. Keeping things the way they are has its merits. It won't be 3 lines though, we would replace all those 9 lines with 1 line:
`return chunkOperation(cipher, cipherOffset, cipherLen, plain, plainOffset, false);`

Reply:
I would prefer to leave it as is; two reasons.. I do think its more readable 'not to have to track down the one-liner function'. I admit this is a matter of taste, so 'can be convinced otherwise'. 

But the second reason, I don't know that we should rely on the JIT necessarily inlining things.. (Anecdotally.. I 'cleaned up' `(Montgomery)IntegerPolynomialP256.mult` with a `for(i = 0:4)` loop that can clearly be unrolled but makes for much more readable code. Never was unrolled)

Reply:
> @ascarpino But a method call should be very cheap, we are adding a bunch of extra implEncrypt/implDecrypt calls here already. Besides, compiler/hotspot will optimize that call if needed. It will be just a method wrapping the `for` loop.

I have done a lot of performance testing with AES/GCM in similar situations by looking at the byte code generated, running JFR, and performance testing with hotspot debugging.  You are right that the "method call should be very cheap", but I've seen code like you describe returning slower results.  This loop is in the hot path for calling the intrinsic and not occasional used.  I will never say absolutely one way or the other, but this looks like a situation where the more verbose syntax will be better.